pkg/ebpf/compiler/compiler.cpp

```diff
@@ -0,0 +1,273 @@
+#include "compiler.h"
+#include "files.h"
+
+#include <iostream>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Support/TargetSelect.h>
+#include <clang/CodeGen/CodeGenAction.h>
+#include <clang/Driver/Job.h>
+#include <clang/Driver/Tool.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Lex/PreprocessorOptions.h>
+
+bool ClangCompiler::llvmInitialized = false;
+
+enum Architecture { PPC, PPCLE, S390X, ARM64, X86 };
+
+std::map<std::string, std::unique_ptr<llvm::MemoryBuffer>> ClangCompiler::remapped_files_;
+
+ClangCompiler::ClangCompiler(const char *name) :
+    llvmContext(new llvm::LLVMContext),
+    diagOpts(new clang::DiagnosticOptions()),
+    errStream(errString),
+    textDiagnosticPrinter(new clang::TextDiagnosticPrinter(errStream, diagOpts.get())),
+    diagnosticsEngine(new clang::DiagnosticsEngine(diagID, diagOpts, textDiagnosticPrinter.get(), false)),
+    defaultCflags({
+        "clang", // DO NOT REMOVE, first flag is ignored
+        "-emit-llvm",
+        "-O2",
+        "-D__KERNEL__",
+        "-fno-color-diagnostics",
+        "-fno-unwind-tables",
+        "-fno-asynchronous-unwind-tables",
+        "-fno-stack-protector",
+        "-nostdinc",
+        "-isystem/virtual/lib/clang/include",
+        "-x", "c"
+    }),
+    theTriple("bpf")
+{
+    if (!ClangCompiler::llvmInitialized) {
+        LLVMInitializeBPFTarget();
+        LLVMInitializeBPFTargetMC();
+        LLVMInitializeBPFTargetInfo();
+        LLVMInitializeBPFAsmPrinter();
+        LLVMInitializeBPFAsmParser();
+
+        ClangCompiler::llvmInitialized = true;
+    }
+
+    theDriver = std::make_unique<clang::driver::Driver>(
+        name, getArch(), *diagnosticsEngine,
+        llvm::vfs::getRealFileSystem()
+    );
+
+    std::string arch = "bpf";
+    std::string Error;
+
+    theTarget = llvm::TargetRegistry::lookupTarget(theTriple.getTriple(), Error);
+    if (!theTarget) {
+        errString = "could not lookup target";
+        return;
+    }
+
+    llvm::TargetOptions targetOptions;
+    auto RM = llvm::Optional<llvm::Reloc::Model>();
+    targetMachine = std::unique_ptr<llvm::TargetMachine>(theTarget->createTargetMachine(
+        theTriple.getTriple(), "generic", "", targetOptions, RM, llvm::None, llvm::CodeGenOpt::Aggressive));
+
+    if (!targetMachine) {
+        errString = "could not allocate target machine";
+        return;
+    }
+
+    if (remapped_files_.empty()) {
+        for (auto f : MappedFiles::files) {
+            remapped_files_[f.first] = llvm::MemoryBuffer::getMemBuffer(f.second);
+        }
+    }
+}
+
+std::unique_ptr<clang::CompilerInvocation> ClangCompiler::buildCompilation(
+    const char *inputFile,
+    const char *outputFile,
+    const std::vector<const char*> &extraCflags,
+    bool verbose)
+{
+    std::vector<const char*> cflags;
+    for (auto it = defaultCflags.begin(); it != defaultCflags.end(); it++)
+        cflags.push_back(*it);
```

Comment:
Can be replaced by `auto cflags = defaultCflags;`

Reply:
Will that create a copy of `defaultCflags`?